pkg/trace/api/profiles.go

```diff
@@ -157,12 +164,148 @@ type multiTransport struct {
 	keys    []string
 }
 
+// isRetryableError determines if an error should trigger a retry
+func isRetryableError(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// Check for network-level errors that indicate connection issues
+	if netErr, ok := err.(net.Error); ok {
+		// Retry on timeout or temporary network errors
+		return netErr.Timeout() || netErr.Temporary()
+	}
+
+	// Check for specific connection errors
+	if opErr, ok := err.(*net.OpError); ok {
+		if opErr.Op == "dial" || opErr.Op == "read" || opErr.Op == "write" {
+			return true
+		}
+		if syscallErr, ok := opErr.Err.(*net.DNSError); ok {
+			return syscallErr.Temporary()
+		}
+		if syscallErr, ok := opErr.Err.(*syscall.Errno); ok {
+			// Common retryable syscall errors
+			switch *syscallErr {
+			case syscall.ECONNREFUSED, syscall.ECONNRESET, syscall.ETIMEDOUT, syscall.ENETUNREACH, syscall.EHOSTUNREACH:
+				return true
+			}
+		}
+	}
+
+	// Check for URL errors that might be retryable
+	if urlErr, ok := err.(*url.Error); ok {
+		return isRetryableError(urlErr.Err)
+	}
+
+	return false
+}
+
+// isRetryableStatusCode determines if an HTTP status code should trigger a retry
+func isRetryableStatusCode(statusCode int) bool {
+	switch statusCode {
+	case http.StatusBadGateway, // 502 - proxy/gateway connection issues
+		http.StatusServiceUnavailable, // 503 - temporary server overload
+		http.StatusGatewayTimeout:     // 504 - proxy/gateway timeout
+		return true
+	default:
+		return false
+	}
+}
+
+// calculateRetryDelay computes the delay for a retry attempt using exponential backoff
+func calculateRetryDelay(attempt int) time.Duration {
+	if attempt <= 0 {
+		return retryBaseDelay
+	}
+
+	delay := retryBaseDelay
+	for i := 0; i < attempt; i++ {
+		delay *= 2
+		if delay > retryMaxDelay {
+			return retryMaxDelay
+		}
+	}
+	return delay
+}
+
 func (m *multiTransport) RoundTrip(req *http.Request) (rresp *http.Response, rerr error) {
 	setTarget := func(r *http.Request, u *url.URL, apiKey string) {
 		r.Host = u.Host
 		r.URL = u
 		r.Header.Set("DD-API-KEY", apiKey)
 	}
+
+	// retryRoundTrip performs a single request with retry logic
+	retryRoundTrip := func(req *http.Request, u *url.URL, apiKey string) (*http.Response, error) {
+		var lastErr error
+		var resp *http.Response
+		var bodyBytes []byte
+
+		// Read the request body once for all retry attempts
+		if req.Body != nil {
+			var err error
+			bodyBytes, err = io.ReadAll(req.Body)
```

Comment:
so reading the body is not great, though it could get consumed by the initial request even if we need to retry